### PR TITLE
Receive API, authentication & authorization

### PR DESCRIPTION
**Pull request by fivitti**
- [x] Receive API in content server
- [x] HMAC authorization
- [x] HMAC DB changes
- [x] HMAC station code
- [x] Remove station-side processing and upload via SFTP by station

### CODE REVIEW

File: backend/app/authorize_station.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import psycopg2
+
+from . import app
+from .hmac_token import parse_token, validate_token
+
+AUTHORIZATION_ALGORITHM = "HMAC-SHA256"
+
+cfg = app.config["database"]
+
+def _get_body(request):
```
Comment (tomaszmrugalski) [inline backend/app/authorize_station.py:15]: The whole file (all files for that matter) lack any comments.

File: backend/app/hmac_token.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    if check_date is None:
+        check_date = datetime.datetime.utcnow()
+    delta = abs(check_date - create_date)
+    if delta > SIG_LIFETIME:
+        return False, id_
+
+    if _verify_signature(sig, secret, body, create_date):
+        return True, id_
+    else:
+        return False, id_
```
Comment (tomaszmrugalski) [inline backend/app/hmac_token.py]: Missing enter on the last line.

File: backend/app/hmac_token.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import datetime
+import hashlib
+import hmac
+from typing import Dict
+
+SIG_LIFETIME = datetime.timedelta(minutes=2, seconds=30)
```
Comment (tomaszmrugalski) [inline backend/app/hmac_token.py:13]: Missing comments in the whole file. At least there should be function comments.

File: backend/app/authorize_station.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    with conn.cursor() as c:
+        query = "SELECT secret FROM stations WHERE station_id = %s"
+        c.execute(query, (station_id,))
+        row = c.fetchone()
+        if row is None:
+            return None
+        return row[0].tobytes()
+
+def _verify_request():
+    header = request.headers.get("Authorization")
```
Comment (tomaszmrugalski) [inline backend/app/authorize_station.py:49]: The return parameters should be documented.

File: backend/app/receive.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                "INSERT INTO observations (aos, tca, los, sat_id, sat_name, filename, notes, station_id)"
+                "VALUES (%(aos)s, %(tca)s, %(los)s, %(sat_id)s, %(sat_name)s, %(filename)s, %(notes)s, %(station_id)s);",
+                {
+                    'aos': args['aos'].isoformat(),
+                    'tca': args['tca'].isoformat(),
+                    'los': args['los'].isoformat(),
+                    'sat_id': sat_id,
+                    'sat_name': args['sat'],
+                    'filename': filename,
+                    'notes': args.get('notes'),
```
Comment (tomaszmrugalski) [inline backend/app/receive.py:50]: Notes are optional. What would happen if they were not provided? args.get('notes') would return None or throw?

File: backend/app/receive.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    'file': fields.Field(validate=lambda file: file.mimetype == "image/png", location="files"),
+    'aos': fields.DateTime(required=True),
+    'tca': fields.DateTime(required=True),
+    'los': fields.DateTime(required=True),
+    'sat': fields.Str(required=True),
+    'notes': fields.Str(required=False)
+})
+def receive(station_id, args):
+    if len(request.files) == 0:
+        abort(400)
```
Comment (tomaszmrugalski) [inline backend/app/receive.py]: There should be an error message communicated back, so the user knows what's wrong.

File: backend/app/receive.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    file_ = args['file']
+    filename = "%s-%s" % (str(uuid.uuid4()), file_.filename)
+
+    with psycopg2.connect(**cfg) as conn:
+        with conn.cursor() as cursor:
+            cursor.execute("SELECT sat_id FROM satellites WHERE sat_name = %s LIMIT 1;", (args["sat"],))
+            row = cursor.fetchone()
+            if row is None:
+                abort(400)
```
Comment (tomaszmrugalski) [inline backend/app/receive.py]: The same as above - error information should be communicated back. Something like "No satellite XYZ in the database.". Also, I think the logic here is broken. In the future the system should accept observations even for unknown sats. So I'd use sat_id -1 or 0 to designate unknown sat and just accept the observation.

File: backend/run_server.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export FLASK_APP=app
```
Comment (tomaszmrugalski) [inline backend/run_server.sh]: There is already satnogs-web.py script. If you don't like and want to replace it - fine. We should have one run script, not two.

File: backend/tests/requests.http [hunk 9]
```diff
@@ -0,0 +1,26 @@
+POST http://127.0.0.1:5000/receive HTTP/1.1
```
Comment (tomaszmrugalski) [inline backend/tests/requests.http]: This looks like part of a test, but there's no actual test using this file. Did you forget to check it in?

Either add missing test files or remove this and x.png file. Obviously, having a test is better than having none.
Note there are 5 unit-tests on master for Flask already. They're very basic, but perhaps you could
rebase the branch and expand on the very basic framework that's already there? See [test_satnogs.py](https://github.com/gut-space/satnogs/blob/master/backend/tests/test_satnogs.py).

File: backend/app/receive.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+from webargs import fields
+from webargs.flaskparser import use_args
+
+from . import app
+from .authorize_station import authorize_station
+from .utils import make_thumbnail
+
+cfg = app.config["database"]
+
+@app.route('/receive', methods=["POST",])
```
Comment (tomaszmrugalski) [inline backend/app/receive.py:16]: An explanation is needed what's the expected syntax is.

**Review by tomaszmrugalski — CHANGES_REQUESTED** [review]

File: backend/app/receive.py (hunk 5 above)
Comment (fivitti) [reply backend/app/receive.py:50]: ```.get('foo')``` return ```None``` by default if ```foo``` key not exists. You can provide second argument with default value. For example ```.get('foo', 42)``` returns 42 when 'foo' key not exists.
```bar['foo']``` throws ```KeyError``` if key not exists

File: backend/tests/requests.http (hunk 9 above)
Comment (fivitti) [reply backend/tests/requests.http]: It isn't test. It is example of HTTP message which should be passed to receive action.  
I use RestClient Visual Studio Code extension to send it. It is simply and may be requests may be commited to repo.
Maybe I should move this file to "request" directory or similar?

File: backend/app/receive.py (hunk 10 above)
Comment (fivitti) [reply backend/app/receive.py:16]: What I should explain here?
```methods``` parameter requires list of available HTTP methods. In Python when list has only one item then after this item should be trailing comma.

File: backend/app/authorize_station.py (hunk 1 above)
Comment (fivitti) [reply backend/app/authorize_station.py:15]: Fix

File: backend/app/hmac_token.py (hunk 3 above)
Comment (fivitti) [reply backend/app/hmac_token.py:13]: Fixed

File: backend/app/authorize_station.py (hunk 4 above)
Comment (fivitti) [reply backend/app/authorize_station.py:49]: Check if my description is enough.

File: backend/app/receive.py (hunk 6 above)
Comment (fivitti) [reply backend/app/receive.py]: Fixed

File: backend/app/receive.py (hunk 7 above)
Comment (fivitti) [reply backend/app/receive.py]: Fixed

File: backend/run_server.sh (hunk 8 above)
Comment (fivitti) [reply backend/run_server.sh]: I deleted my version.

File: backend/app/hmac_token.py (hunk 2 above)
Comment (fivitti) [reply backend/app/hmac_token.py]: Fixed

File: backend/app/receive.py (hunk 10 above)
Comment (tomaszmrugalski) [reply backend/app/receive.py:16]: My comment was about the whole /receive point. If I want to use this API, I should send a POST command. But how do I pass the parameters? In http headers? If so, what are the names?

File: backend/app/receive.py (hunk 5 above)
Comment (tomaszmrugalski) [reply backend/app/receive.py:50]: Cool trick. Thanks!

File: backend/tests/requests.http (hunk 9 above)
Comment (tomaszmrugalski) [reply backend/tests/requests.http]: Moved to backend/examples/

File: backend/app/authorize_station.py (hunk 4 above)
Comment (tomaszmrugalski) [reply backend/app/authorize_station.py:49]: I've pushed updated comment.

File: backend/app/receive.py (hunk 10 above)
Comment (tomaszmrugalski) [reply backend/app/receive.py:16]: I saw the requests.http file. This is enough for explanation.

**Review by tomaszmrugalski — APPROVED** [review]

**Pull request merged**